engine/chassis.ts

```diff
@@ -88,17 +212,218 @@ function hashObject (o: object) {
+                bindings.push(new DescriptorSetLayoutBinding(
+                    bindings.length,
+                    descriptorType,
+                    1,
+                    shaderStageFlag,
+                ));
+            }
+        }
+
+        for (const samplerTexture of info.samplerTextures || []) {
+            const bindings = layoutBindings[samplerTexture.set];
+            const regex = new RegExp(`${reg}${samplerTexture.name}${reg}`, `g`);
+            let shaderStageFlag = ShaderStageFlagBit.NONE;
+            if (info.frag.search(regex)) shaderStageFlag |= ShaderStageFlagBit.FRAGMENT;
+            if (info.vert.search(regex)) shaderStageFlag |= ShaderStageFlagBit.VERTEX;
+
+            if (shaderStageFlag === ShaderStageFlagBit.NONE) continue;
+            samplerTextures.push(new UniformSamplerTexture(
+                samplerTexture.set,
+                bindings.length,
+                samplerTexture.name,
+                samplerTexture.type,
+                samplerTexture.count || 1,
+            ));
+            // immutableSampler?
+            bindings.push(new DescriptorSetLayoutBinding(
+                bindings.length,
+                DescriptorType.SAMPLER_TEXTURE,
+                1,
+                shaderStageFlag,
+            ));
+        }
+
+        for (const block of shaderBlocks || []) {
+            let offset = 0;
+            for (const member of block.members) {
+                this._propertyMap[member.name] = genHandle(
+                    block.set,
+                    block.binding,
+                    member.type,
+                    member.count || 1,
+                    offset,
+                );
+                TestBase.assert(GetTypeSize(member.type) >= GetTypeSize(Type.UINT4) || (member.count || 1) <= 1,
+                    `the width of ${TypeInfos[member.type].glslName} will be like vec4 in glsl std140, don't use array`);
+                // convert unit to int/float/uint
+                offset += (TypeInfos[member.type].size >> 2) * (member.count || 1);
+            }
+        }
+
+        for (const samplerTexture of samplerTextures) {
+            this._propertyMap[samplerTexture.name] = genHandle(
+                samplerTexture.set,
+                samplerTexture.binding,
+                samplerTexture.type,
+                samplerTexture.count || 1,
+            );
+        }
+
+        for (const descriptorSetId of descriptorSetIds) {
+            const blocks = shaderBlocks?.filter((block) => block.set === descriptorSetId) || [];
+            const textures = samplerTextures?.filter((texture) => texture.set === descriptorSetId) || [];
+
+            TestBase.assert(blocks.length < TEST_UBO_COUNTS[descriptorSetId] || TEST_UBO_COUNTS[descriptorSetId] === -1,
+                `too many UBOs in set ${descriptorSetId}, only ${TEST_UBO_COUNTS[descriptorSetId]} seats exist.`);
+            TestBase.assert(blocks.length < 8 && TEST_UBO_COUNTS[descriptorSetId] !== -1,
```

Comment:
Do not use constants that are not immediately clear where they come from, just write out the calculation process